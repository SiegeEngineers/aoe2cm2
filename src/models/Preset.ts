import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";
import Civilisation from "./Civilisation";
import DraftOption from "./DraftOption";
import {Util} from "../util/Util";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);

    public readonly name: string;
    public presetId?: string
    public readonly encodedCivilisations?: string;
    public readonly draftOptions?: DraftOption[];
    public readonly turns: Turn[];

    constructor(name: string, draftOptions: DraftOption[], turns: Turn[] = [], presetId?: string) {
        this.name = name;
        this.presetId = presetId;
        if (Util.isCivilisationArray(draftOptions)) {
            this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(draftOptions);
        } else {
            this.draftOptions = draftOptions;
        }
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations?: string, draftOptions?: DraftOption[], turns: Turn[], presetId?: string } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isOptionalString(preset.encodedCivilisations);
        Assert.isOptionalString(preset.presetId);
        let draftOptions: DraftOption[] = [];
        if (preset.encodedCivilisations) {
            draftOptions = CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations);
        } else if (preset.draftOptions) {
            draftOptions = preset.draftOptions;
        }
        return new Preset(preset.name, draftOptions, Turn.fromPojoArray(preset.turns), preset.presetId);
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get options(): DraftOption[] {
        if (this.encodedCivilisations) {
            return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
        }
        if (this.draftOptions) {
            return this.draftOptions;
        }
        throw new Error('Invalid Preset without either encodedCivilisations or draftOptions');
    }
}

export default Preset;