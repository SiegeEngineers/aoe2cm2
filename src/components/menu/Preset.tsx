import * as React from "react";
import {PresetUtil} from "../../util/PresetUtil";
import {default as ModelPreset} from "../../models/Preset"
import TurnRow from "../draft/TurnRow";
import NewDraftButton from "../NewDraftButton";
import CustomisePresetButton from "../../containers/CustomisePresetButton";
import CopyableInput from "../draft/CopyableInput";
import {PresetOptionCheckbox} from "../PresetEditor/PresetOptionCheckbox";
import Civilisation from "../../models/Civilisation";
import DraftOptionPanelType from "../../constants/DraftOptionPanelType";
import DraftOptionPanel from "../draft/DraftOptionPanel";
import DraftOption from "../../models/DraftOption";
import {ApplicationState} from "../../types";
import {Dispatch} from "redux";
import * as actions from "../../actions";
import {connect} from "react-redux";

interface IState {
    preset?: ModelPreset;
    presetExists: boolean;
}

interface IProps {
    iconStyle: string;
}

class Preset extends React.Component<IProps, IState> {

    state: IState = {presetExists: true};

    componentDidMount() {
        const presetId = PresetUtil.getIdFromUrl();
        if (presetId !== undefined) {
            this.loadPreset(presetId);
        }
    }

    public render() {
        if (!this.state.presetExists) {
            return (
                <div className="content box">
                    <p>Could not find this preset, sorry :-(</p>
                </div>
            );
        }
        if (this.state.preset !== undefined) {

            const presetCivilisations = this.state.preset.options;
            let civs;
            let itemAlignment = '';
            if (this.state.preset.encodedCivilisations) {
                civs = Civilisation.ALL.map((value: Civilisation, index: number) =>
                    <PresetOptionCheckbox presetOptions={presetCivilisations} value={value}
                                          key={index}
                                          disabled={true}/>)
            } else {
                civs = presetCivilisations.map((value: DraftOption, index: number) =>
                    <DraftOptionPanel draftOption={value} active={false}
                                      draftOptionPanelType={DraftOptionPanelType.CHOICE} nextAction={0}
                                      displayOnly={true} iconStyle={this.props.iconStyle}/>);
                itemAlignment = ' flex-justify-center';
            }

            return (
                <div className='content box'>
                    <h3 className="has-text-centered">{this.state.preset.name}</h3>

                    <TurnRow turns={this.state.preset.turns}/>

                    <div className={"is-flex" + itemAlignment} style={{flexDirection: 'row', flexWrap: 'wrap'}}>
                        {civs}
                    </div>

                    <div className="columns is-mobile mt-4">
                        <div className="column is-7 buttons">
                            <br/>
                            <NewDraftButton preset={this.state.preset}/>
                            <CustomisePresetButton preset={this.state.preset}/>
                        </div>
                        <div className="column is-5">
                            <CopyableInput content={window.location.href} before={'preset.shareThisPreset'}
                                           length={35}/>
                        </div>
                    </div>
                </div>
            );
        }
        return (
            <div className="content box">
                <p>Loading…</p>
            </div>
        );
    }


    private loadPreset = (presetId: string) => {
        const request = new XMLHttpRequest();
        request.open('GET', '/api/preset/' + presetId, true);
        request.onreadystatechange = () => {
            if (request.readyState === XMLHttpRequest.DONE && request.status === 200) {
                const result = JSON.parse(request.responseText);
                const preset = ModelPreset.fromPojo(result);
                this.setState({preset});
                document.title = `Preset "${preset?.name}" – AoE2 Captains Mode`;
            } else if (request.readyState === XMLHttpRequest.DONE && request.status === 404) {
                this.setState({presetExists: false});
                document.title = 'Preset not found – AoE2 Captains Mode';
            }
        };
        request.send();
    };
}


function mapStateToProps(state: ApplicationState) {
    return {
        iconStyle: state.iconStyle.iconStyle,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(Preset);
