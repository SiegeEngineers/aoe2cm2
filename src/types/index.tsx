import Preset from "../models/Preset";
import Player from "../constants/Player";
import {DraftEvent} from "./DraftEvent";
import {ColorScheme} from "../constants/ColorScheme";

export interface IDraftState {
    nameHost: string;
    nameGuest: string;
    hostConnected: boolean;
    guestConnected: boolean;
    hostReady: boolean;
    guestReady: boolean;
    preset?: Preset;
    events: DraftEvent[];
    startTimestamp: number;
}

export interface IReplayState {
    events: DraftEvent[];
    countdownInterval: NodeJS.Timeout | null;
    stopCountdown: NodeJS.Timeout | null;
    eventTimeouts: NodeJS.Timeout[];
}

export interface ApplicationState {
    draft: IDraftState,
    countdown: ICountdownState,
    replay: IReplayState,
    ownProperties: IDraftOwnPropertiesState,
    language: ILanguageState,
    iconStyle: IIconStyleState,
    modal: IModalState,
    presetEditor: IPresetEditorState,
    colorScheme: IColorSchemeState
}

export interface IDraftOwnPropertiesState {
    whoAmI?: Player;
    ownName: string | null;
    nextAction: number;
}

export interface ICountdownState {
    countdownValue: number;
    countdownVisible: boolean;
}

export interface IModalState {
    showModal: boolean;
    showRoleModal: boolean;
}

export interface IPresetEditorState {
    editorPreset: Preset | null;
}

export interface ILanguageState {
    language: string;
}

export interface IIconStyleState {
    iconStyle: string;
}


export interface IColorSchemeState {
    colorScheme: ColorScheme;
}

export interface ICountdownValues {
    value: number;
    display: boolean;
}

export interface IRecentDraft {
    title: string;
    draftId: string;
    nameHost: string;
    nameGuest: string;
    ongoing: boolean;
}

export interface IAlert {
    class: string;
    title: string;
    content: string;
    closable: boolean;
}

export interface IServerState {
    maintenanceMode: boolean
    hiddenPresetIds: string[]
}
