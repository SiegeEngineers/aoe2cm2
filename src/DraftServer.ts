import express, {Express, Response as ExpressResponse} from "express";
import {Server} from "http"
import socketio from "socket.io";
import Player from "./constants/Player";
import {IDraftConfig} from "./types/IDraftConfig";
import {ISetRoleMessage} from "./types/ISetRoleMessage";
import {DraftsStore} from "./models/DraftsStore";
import {Validator} from "./models/Validator";
import {ValidationId} from "./constants/ValidationId";
import {DraftEvent} from "./types/DraftEvent";
import {Util} from "./util/Util";
import {AddressInfo} from "net";
import Preset from "./models/Preset";
import {ActListener} from "./util/ActListener";
import * as fs from "fs";
import {logger} from "./util/Logger";
import {ISetNameMessage} from "./types/ISetNameMessage";
import {PresetUtil} from "./util/PresetUtil";
import {IServerState} from "./types";
import path from "path";

const ONE_HOUR = 1000 * 60 * 60;


export class DraftServer {
    readonly serverStateFile: string;
    readonly baseDir: string;
    readonly currentDataDirectory: string;
    readonly dataDirectory: string;
    readonly presetDirectory: string;

    constructor(baseDir: string = './') {
        this.serverStateFile = path.join(baseDir, 'serverState.json');
        this.baseDir = baseDir;
        this.dataDirectory = path.join(baseDir, 'data');
        this.currentDataDirectory = path.join(this.dataDirectory, 'current');
        this.presetDirectory = path.join(baseDir, 'presets');
    }

    private loadState(): IServerState {
        if (fs.existsSync(this.serverStateFile)) {
            const fileContent = fs.readFileSync(this.serverStateFile);
            return JSON.parse(fileContent.toString()) as IServerState;
        } else {
            return {maintenanceMode: false, hiddenPresetIds: []};
        }
    }

    private saveState(state: IServerState) {
        fs.writeFileSync(this.serverStateFile, JSON.stringify(state));
    }

    static plain404 = (res: ExpressResponse<any>) => (err: Error) => {
        if (err) {
            res.status(404).send("Not found.")
        }
    };

    serve(port: string | number | undefined): { httpServerAddr: AddressInfo | string | null; io: SocketIO.Server; httpServer: Server } {
        logger.info("Starting DraftServer on port %d", port);
        const app = express();
        app.set("port", port);
        app.use(express.json());

        const server = new Server(app);
        const io = socketio(server, {cookie: false});
        const state: IServerState = this.loadState();
        const draftsStore = new DraftsStore(this.baseDir, state);

        this.setUpApiRoutes(app, state, draftsStore, io);
        this.setUpSocketIo(io, draftsStore);

        const httpServer = server.listen(port, () => {
            logger.info("Listening on: %d", port);
        });
        const httpServerAddr = httpServer.address();

        setInterval(() => {
            draftsStore.purgeStaleDrafts()
        }, ONE_HOUR);

        return {httpServer, httpServerAddr, io};
    }

    private setUpSocketIo(io: SocketIO.Server, draftsStore: DraftsStore) {
        io.on("connection", (socket: socketio.Socket) => {
            const draftIdRaw: string = socket.handshake.query.draftId;
            const draftId = Util.sanitizeDraftId(draftIdRaw);
            logger.info('Connection for draftId: %s', draftId, {draftId});

            const roomHost: string = `${draftId}-host`;
            const roomGuest: string = `${draftId}-guest`;
            const roomSpec: string = `${draftId}-spec`;

            if (!draftsStore.has(draftId)) {
                try {
                    const draftPath = path.join(this.currentDataDirectory, `${draftId}.json`);
                    if (fs.existsSync(draftPath)) {
                        logger.info("Found recorded draft. Sending replay.", {draftId});
                        socket.emit('replay', JSON.parse(fs.readFileSync(draftPath).toString('utf8')));
                    } else if (draftsStore.hasArchive(draftId)) {
                        logger.info("Found archived draft. Sending replay.", {draftId});
                        const archiveFolder = draftsStore.getArchiveFolder(draftId);
                        const archivedDraftPath = path.join(this.dataDirectory, archiveFolder, `${draftId}.json`);
                        socket.emit('replay', JSON.parse(fs.readFileSync(archivedDraftPath).toString('utf8')));
                    } else {
                        logger.info("No recorded draft found.", {draftId});
                        socket.emit('message', 'This draft does not exist.');
                    }
                } catch (e) {
                    logger.error('Sending replay file failed', {draftId});
                    logger.error(e.message, {draftId});
                }
                logger.info("Disconnecting.", {draftId});
                socket.disconnect(true);
                return;
            }

            const rooms = Object.keys(socket.rooms);
            logger.debug("rooms: %s", JSON.stringify(rooms), {draftId});
            let yourPlayerType = Player.NONE;
            if (rooms.includes(roomHost)) {
                socket.join(roomHost); // async
                yourPlayerType = Player.HOST;
            } else if (rooms.includes(roomGuest)) {
                socket.join(roomGuest); // async
                yourPlayerType = Player.GUEST;
            } else {
                socket.join(roomSpec); // async
            }

            socket.on("set_role", (message: ISetRoleMessage, fn: (dc: IDraftConfig) => void) => {
                logger.info("Player wants to set own role: %s", JSON.stringify(message), {draftId});
                if (!draftsStore.has(draftId)) {
                    logger.warn("Draft does not exist", {draftId});
                    socket.emit('message', 'This draft does not exist.');
                    return;
                }
                const role: Player = Util.sanitizeRole(message.role);
                let assignedRole = Util.getAssignedRole(socket, roomHost, roomGuest);
                const rooms = Object.keys(socket.rooms);
                if (rooms.includes(roomSpec)) {
                    if (role === Player.HOST && !draftsStore.isPlayerConnected(draftId, role)) {
                        logger.info("Setting player role to 'HOST': %s", message.name, {draftId});
                        socket.join(roomHost); // async
                        socket.leave(roomSpec); // async
                        draftsStore.connectPlayer(draftId, Player.HOST, message.name);
                        assignedRole = Player.HOST;
                    } else if (role === Player.GUEST && !draftsStore.isPlayerConnected(draftId, role)) {
                        logger.info("Setting player role to 'GUEST': %s", message.name, {draftId});
                        socket.join(roomGuest); // async
                        socket.leave(roomSpec); // async
                        draftsStore.connectPlayer(draftId, Player.GUEST, message.name);
                        assignedRole = Player.GUEST;
                    } else {
                        logger.info("Setting role not possible: %s", role, {draftId});
                    }
                } else {
                    logger.info("Player is not registered as spectator currently. No action taken.", {draftId});
                }

                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_set_role", {name: message.name, playerType: assignedRole});
                fn({
                    ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(assignedRole),
                    yourPlayerType: assignedRole,
                });
            });

            socket.on("set_name", (message: ISetNameMessage, fn: () => void) => {
                logger.info("Player wants to set own name: %s", JSON.stringify(message), {draftId});
                if (!draftsStore.has(draftId)) {
                    logger.warn("Draft does not exist", {draftId});
                    socket.emit('message', 'This draft does not exist.');
                    return;
                }
                let assignedRole = Util.getAssignedRole(socket, roomHost, roomGuest);
                if (assignedRole === Player.HOST) {
                    logger.info("Setting HOST player name to: %s", message.name, {draftId});
                    draftsStore.setPlayerName(draftId, Player.HOST, message.name);
                } else if (assignedRole === Player.GUEST) {
                    logger.info("Setting GUEST player name to: %s", message.name, {draftId});
                    draftsStore.setPlayerName(draftId, Player.GUEST, message.name);
                } else {
                    logger.info("Player is not registered as HOST or GUEST currently. No action taken.", {draftId});
                    return;
                }

                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_set_name", {name: message.name, playerType: assignedRole});
                fn();
            });

            socket.on("ready", (message: {}, fn: (dc: IDraftConfig) => void) => {
                if (!draftsStore.has(draftId)) {
                    logger.warn("Player wants to indicate readyness, but Draft does not exist", {draftId});
                    socket.emit('message', 'This draft does not exist.');
                    return;
                }
                let assignedRole: Player = Player.NONE;
                let wasAlreadyReady = false;
                if (Object.keys(socket.rooms).includes(roomHost)) {
                    wasAlreadyReady = draftsStore.setPlayerReady(draftId, Player.HOST);
                    assignedRole = Player.HOST;
                } else if (Object.keys(socket.rooms).includes(roomGuest)) {
                    wasAlreadyReady = draftsStore.setPlayerReady(draftId, Player.GUEST);
                    assignedRole = Player.GUEST;
                }
                logger.info("Player indicates they are ready: %s", assignedRole, {draftId});
                if (!wasAlreadyReady && draftsStore.playersAreReady(draftId)) {
                    logger.info("Both Players are ready, starting countdown.", {draftId});
                    draftsStore.startCountdown(draftId, socket, this.currentDataDirectory);
                    draftsStore.setStartTimestamp(draftId);
                }
                socket.nsp
                    .in(roomHost)
                    .in(roomGuest)
                    .in(roomSpec)
                    .emit("player_ready", {playerType: assignedRole});
                fn({
                    ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(assignedRole),
                    yourPlayerType: assignedRole
                });
            });

            const validator = new Validator(draftsStore);

            function validateAndApply(draftId: string, message: DraftEvent): ValidationId[] {
                return validator.validateAndApply(draftId, message);
            }

            socket.on("act", new ActListener(this.currentDataDirectory).actListener(draftsStore, draftId, validateAndApply, socket, roomHost, roomGuest, roomSpec));

            socket.on('disconnecting', function () {
                const assignedRole = Util.getAssignedRole(socket, roomHost, roomGuest);
                logger.info("Player disconnected: %s", assignedRole, {draftId});
                if (!draftsStore.has(draftId)) {
                    return;
                }
                draftsStore.disconnectPlayer(draftId, assignedRole);
            });

            let draftState = {
                ...draftsStore.getDraftViewsOrThrow(draftId).getDraftForPlayer(Player.NONE),
                yourPlayerType: yourPlayerType
            };
            logger.info("Emitting draft_state: %s", JSON.stringify(draftState), {draftId});
            socket.emit("draft_state", draftState);
        });
    }

    private setUpApiRoutes(app: Express, state: IServerState, draftsStore: DraftsStore, io: SocketIO.Server) {
        app.post('/api/state', (req, res) => {
            if (!Util.isRequestFromLocalhost(req)) {
                res.status(403).end();
                return;
            }
            logger.info('Received request to set state mode: %s', JSON.stringify(req.body));
            for (let key in req.body) {
                if (state.hasOwnProperty(key) && req.body.hasOwnProperty(key)) {
                    state[key] = req.body[key];
                }
            }
            res.json(state);
            logger.info('New state = %s', JSON.stringify(state));
            this.saveState(state);
            logger.info('Persisted new state');
        });
        app.get('/api/state', (req, res) => {
            if (!Util.isRequestFromLocalhost(req)) {
                res.status(403).end();
                return;
            }
            res.json(state);
        });
        app.post('/api/draft/new', (req, res) => {
            logger.info('Received request to create a new draft: %s', JSON.stringify(req.body));

            if (state.maintenanceMode) {
                res.json({
                    status: 'error',
                    message: 'aoe2cm is currently in maintenance mode, no drafts can be created'
                });
                logger.info('Server is in maintenance mode. Discarding draft creation request.');
                return;
            }

            let draftId = Util.newDraftId();
            while (draftsStore.draftIdExists(draftId)) {
                draftId += Util.randomChar();
            }
            const pojo: Preset = req.body.preset as Preset;
            let preset = Preset.fromPojo(pojo);
            const validationErrors = Validator.validatePreset(preset);
            if (validationErrors.length === 0) {
                draftsStore.initDraft(draftId, preset as Preset);
                res.json({status: 'ok', draftId});
                logger.info('Created new draft with id: %s', draftId, {draftId});
            } else {
                res.json({status: 'error', validationErrors});
                logger.info('Draft validation failed: %s', JSON.stringify(validationErrors));
            }
        });
        app.post('/api/preset/new', (req, res) => {
            logger.info('Received request to create a new preset: %s', JSON.stringify(req.body));
            const pojo: Preset = req.body.preset as Preset;
            let preset = Preset.fromPojo(pojo);
            const validationErrors = Validator.validatePreset(preset);
            if (validationErrors.length === 0) {
                const presetId = PresetUtil.createPreset(preset as Preset, this.presetDirectory);
                res.json({status: 'ok', presetId: presetId});
                logger.info('Created new preset with id: %s', presetId, {presetId});
            } else {
                res.json({status: 'error', validationErrors});
                logger.info('Preset validation failed: %s', JSON.stringify(validationErrors));
            }
        });
        app.get('/api/connections', (req, res) => {
            if (Util.isRequestFromLocalhost(req)) {
                // @ts-ignore
                res.json({connections: io.engine.clientsCount, rooms: io.sockets.adapter.rooms});
            } else {
                // @ts-ignore
                res.json({connections: io.engine.clientsCount});
            }
        });
        app.get('/api/alerts', (req, res) => {
            res.sendFile('alerts.json', {'root': __dirname + '/..'});
        });
        app.get('/api/preset/list', (req, res) => {
            res.sendFile('presets.json', {'root': __dirname + '/..'});
        });
        app.get('/api/preset/:id', (req, res) => {
            res.sendFile(req.params.id + '.json', {'root': __dirname + '/../presets'}, DraftServer.plain404(res));
        });
        app.get('/api/recentdrafts', (req, res) => {
            res.json(draftsStore.getRecentDrafts());
        });
        app.get('/api/draft/:id', (req, res) => {
            const draftId = req.params.id;
            if (draftsStore.hasArchive(draftId)) {
                const archiveDirectory = draftsStore.getArchiveFolder(draftId);
                res.sendFile(draftId + '.json', {'root': path.join(this.dataDirectory, archiveDirectory)}, DraftServer.plain404(res));
            } else {
                res.sendFile(draftId + '.json', {'root': this.currentDataDirectory}, DraftServer.plain404(res));
            }
        });

        const indexPath = __dirname + '/index.html';

        app.use('/draft/[a-zA-Z]+', (req, res) => {
            if (fs.existsSync(indexPath)) {
                res.sendFile(indexPath);
            } else {
                res.sendFile('maintenance.html', {'root': __dirname + '/../public'});
            }
        });

        app.use(express.static('build'));
        app.use('/', (req, res) => {
            if (fs.existsSync(indexPath)) {
                res.sendFile(indexPath);
            } else {
                res.sendFile('maintenance.html', {'root': __dirname + '/../public'});
            }
        });
    }
}
