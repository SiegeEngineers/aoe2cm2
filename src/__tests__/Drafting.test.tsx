import {default as io} from "socket.io-client"
import {DraftServer} from "../DraftServer";
import request from "request"
import Player from "../constants/Player";
import {IDraftConfig} from "../types/IDraftConfig";
import getPort from "get-port";
import Preset from "../models/Preset";
import {Barrier} from "../test/Barrier";
import temp from "temp";
import * as fs from "fs";
import path from "path";

let hostSocket: any;
let clientSocket: any;
let spectatorSocket: any;
let httpServer: any;
let httpServerAddr: any;
let ioServer: any;
let draftId: string;
let draftServer: DraftServer;


beforeAll(() => {
    temp.track();
    const dirPath = temp.mkdirSync('serverTest');
    fs.mkdirSync(path.join(dirPath, 'data', 'current'), {recursive: true});
    draftServer = new DraftServer(dirPath);
});

afterAll(() => {
    temp.cleanupSync();
});


beforeAll((done) => {
    getPort().then((port: number) => {
        console.log("Got port: " + port);
        const serve = draftServer.serve(port);
        httpServer = serve.httpServer;
        httpServerAddr = serve.httpServerAddr;
        ioServer = serve.io;
        done();
    });
});

afterAll((done) => {
    ioServer.close();
    httpServer.close();
    done();
});

function connect() {
    return io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
        query: {draftId: draftId},
        reconnectionDelay: 0,
        forceNew: true,
        transports: ['websocket'],
    });
}

beforeEach((done) => {
    const barrier = new Barrier(3, done);
    request.post(`http://[${httpServerAddr.address}]:${httpServerAddr.port}/api/draft/new`,
        {body: JSON.stringify({preset: Preset.SIMPLE}), headers: {'Content-Type': 'application/json; charset=UTF-8'}},
        (error, response, body) => {
            const draftIdContainer: { draftId: string } = JSON.parse(body);
            draftId = draftIdContainer.draftId;

            hostSocket = connect();
            hostSocket.on('connect', () => {
                barrier.trigger();
            });

            clientSocket = connect();
            clientSocket.on('connect', () => {
                barrier.trigger();
            });

            spectatorSocket = connect();
            spectatorSocket.on('connect', () => {
                barrier.trigger();
            });
        });
});

afterEach((done) => {
    if (hostSocket.connected) {
        hostSocket.disconnect();
    }
    if (clientSocket.connected) {
        clientSocket.disconnect();
    }
    if (spectatorSocket.connected) {
        spectatorSocket.disconnect();
    }
    done();
});

it('successful join gets a draft config', (done) => {
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, (data: IDraftConfig) => {
        data.startTimestamp = 0;
        expect(data).toMatchSnapshot();
        done();
    });
});


it('should send player_set_role when player sets role', (done) => {
    const barrier = new Barrier(2, done);
    hostSocket.once("player_set_role", (message: any) => {
        expect(message.name).toBe('Saladin');
        expect(message.playerType).toBe(Player.HOST);

        hostSocket.once("player_set_role", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
            barrier.trigger();
        });
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
        });
    });

    clientSocket.once("player_set_role", (message: any) => {
        expect(message.name).toBe('Saladin');
        expect(message.playerType).toBe(Player.HOST);

        clientSocket.once("player_set_role", (message: any) => {
            expect(message.name).toBe('Barbarossa');
            expect(message.playerType).toBe(Player.GUEST);
            barrier.trigger();
        });
    });

    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
    });
});

it('players can change their names', (done) => {
    const barrier = new Barrier(4, done);
    hostSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        hostSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    clientSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        clientSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    spectatorSocket.once('player_set_name', (message: any) => {
        expect(message.name).toEqual('Attila The Hun');
        spectatorSocket.once('player_set_name', (message: any) => {
            expect(message.name).toEqual('Bleda The Hun');
            barrier.trigger();
        });
    });
    const newHostName = "Attila The Hun";
    const newGuestName = "Bleda The Hun";
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
            clientSocket.emit('ready', {}, () => {
                hostSocket.emit('set_name', {
                    "name": newHostName
                }, () => {
                    hostSocket.emit('ready', {}, () => {
                        clientSocket.emit('set_name', {
                            "name": newGuestName
                        }, () => {
                            // done
                            const secondSpectatorSocket = io.connect(`http://[${httpServerAddr.address}]:${httpServerAddr.port}`, {
                                query: {draftId: draftId},
                                reconnectionDelay: 0,
                                forceNew: true,
                                transports: ['websocket'],
                            });
                            secondSpectatorSocket.once('draft_state', (message: any) => {
                                expect(message.nameHost).toEqual(newHostName);
                                expect(message.nameGuest).toEqual(newGuestName);
                                barrier.trigger();
                            });
                        });
                    });
                });
            });
        });
    });
});

it('fully execute sample draft', (done) => {
    const barrier = new Barrier(2, done);
    hostSocket.once('disconnect', () => {
        barrier.trigger();
    });
    clientSocket.once('disconnect', () => {
        barrier.trigger();
    });
    hostSocket.emit('set_role', {name: 'Saladin', role: Player.HOST}, () => {
        clientSocket.emit('set_role', {name: 'Barbarossa', role: Player.GUEST}, () => {
            clientSocket.emit('ready', {}, () => {
                hostSocket.emit('ready', {}, () => {
                    hostSocket.emit('act', {
                        "player": "HOST",
                        "executingPlayer": "HOST",
                        "actionType": "ban",
                        "chosenOptionId": "Celts",
                        "isRandomlyChosen": false,
                    }, () => {
                        clientSocket.emit('act', {
                            "player": "GUEST",
                            "executingPlayer": "GUEST",
                            "actionType": "ban",
                            "chosenOptionId": "Celts",
                            "isRandomlyChosen": false,
                        }, () => {
                            clientSocket.emit('act', {
                                "player": "GUEST",
                                "executingPlayer": "GUEST",
                                "actionType": "pick",
                                "chosenOptionId": "Slavs",
                                "isRandomlyChosen": false,
                            }, () => {
                                hostSocket.emit('act', {
                                    "player": "HOST",
                                    "executingPlayer": "HOST",
                                    "actionType": "pick",
                                    "chosenOptionId": "Slavs",
                                    "isRandomlyChosen": false,
                                }, () => {
                                    // preset done
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});