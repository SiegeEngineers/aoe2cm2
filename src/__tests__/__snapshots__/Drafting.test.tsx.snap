// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful join gets a draft config 1`] = `
Object {
  "events": Array [],
  "guestConnected": false,
  "guestReady": false,
  "hostConnected": true,
  "hostReady": false,
  "nameGuest": "â€¦",
  "nameHost": "Saladin",
  "nextAction": 0,
  "preset": Object {
    "encodedCivilisations": "7fffffffff",
    "name": "Simple Preset",
    "turns": Array [
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "BAN",
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "GUEST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "NONEXCLUSIVE",
        "executingPlayer": "HOST",
        "hidden": false,
        "id": "mocked-uuid",
        "parallel": false,
        "player": "HOST",
      },
    ],
  },
  "startTimestamp": 0,
  "yourPlayerType": "HOST",
}
`;
